src/lib.rs

```diff
@@ -311,123 +214,179 @@ impl Client {
     /// server at all.
     ///
     /// Available as of memcached 1.4.31.
-    pub async fn dump_keys(&mut self) -> Result<MetadumpIter<'_>, Error> {
-        self.conn.write_all(b"lru_crawler metadump all\r\n").await?;
-        self.conn.flush().await?;
-
-        Ok(MetadumpIter {
-            client: self,
-            done: false,
-        })
+    pub async fn dump_keys(&mut self) -> Result<Vec<MetadumpIter<'_>>, Error> {
+        let result = stream::iter(&mut self.ring.servers)
+            .then(|node| node.dump_keys())
+            .collect::<KeyDumpResult>()
+            .await;
+
+        // Return the result from the KeyDumpResult new type struct
+        result.0
     }
 
     /// Collects statistics from the server.
     ///
     /// The statistics that may be returned are detailed in the protocol specification for
     /// memcached, but all values returned by this method are returned as strings and are not
     /// further interpreted or validated for conformity.
-    pub async fn stats(&mut self) -> Result<HashMap<String, String>, Error> {
-        let mut entries = HashMap::new();
-
-        self.conn.write_all(b"stats\r\n").await?;
-        self.conn.flush().await?;
+    pub async fn stats(&mut self) -> Result<Vec<HashMap<String, String>>, Error> {
+        let result = stream::iter(&mut self.ring.servers)
+            .then(|node| node.stats())
+            .collect::<StatsResult>()
+            .await;
+
+        // Return the result from the StatsResult new type struct
+        result.0
+    }
+}
 
-        while let StatsResponse::Entry(key, value) = self.get_stats_response().await? {
-            entries.insert(key, value);
-        }
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::{BufRead, Write};
+    use std::net::{TcpListener, TcpStream};
 
-        Ok(entries)
+    const KEY: &str = "async-memcache-test-key";
+    const EMPTY_KEY: &str = "no-value-here";
+    const SERVER_ADDRESSES: [&str; 2] = ["localhost:1234", "localhost:1235"];
+
+    #[ctor::ctor]
+    fn init() {
+        SERVER_ADDRESSES.iter().for_each(|dsn| {
+            let listener = TcpListener::bind(dsn).expect("Failed to bind listener");
+
+            // accept connections and process them serially in a background thread
+            std::thread::spawn(move || {
+                for stream in listener.incoming() {
+                    let stream = stream.expect("Failed to accept connection");
+                    std::thread::spawn(move || {
+                        handle_connection(stream);
+                    });
+                }
+            });
+        });
     }
```

Comment:
This feels very integration-test-like. Should we move to `tests/integration_test.rs` or something like that?